lib/internal/util.js

```diff
@@ -187,3 +191,65 @@ exports.convertToValidSignal = function convertToValidSignal(signal) {
 
   throw new Error('Unknown signal: ' + signal);
 };
+
+const kCustomPromisifiedSymbol = Symbol('util.promisify.custom');
+const kCustomPromisifyArgsSymbol = Symbol('customPromisifyArgs');
+
+function promisify(orig) {
+  if (typeof orig !== 'function') {
+    throw new TypeError('The first argument to promisify() must be a function');
+  }
+
+  if (orig[kCustomPromisifiedSymbol]) {
+    const fn = orig[kCustomPromisifiedSymbol];
+    Object.defineProperty(fn, kCustomPromisifiedSymbol, {
+      value: fn, enumerable: false, writable: false, configurable: true
+    });
+    return fn;
+  }
+
+  if (isAsyncFunction(orig)) {
```

Comment:
In my opinion, it would be confusing to have special behavior for async functions. As a user, I would expect to be able to convert any Promise-returning function to an analogous async function without worrying about behavior changes. This is usually a safe assumption because async functions are indistinguishable from regular functions*, but it won't hold if `util.promisify` has "magic" behavior and treats async functions differently.

My intuitive model of this function is that it should behave like this (with better performance):

```js
function promisify(func) {
  return (...args) =>
    new Promise((resolve, reject) =>
      func(...args, (err, result) => err ? reject(err) : resolve(result))
    );
}
```

This is a relatively simple spec and is easy to understand. I think `util.promisify` will get more confusing the more it deviates from this.

<sup>*Aside from rarely-used operations such as `func.toString()` and `func.constructor`</sup>